text/0005-application-formalization.md

```diff
@@ -0,0 +1,196 @@
+  * PUBLIC REVIEW: https://github.com/folio-org/rfcs/pull/29
+  * FINAL REVIEW: TBD
+* Outcome: <!-- Leave this blank. Will eventually be either ACCEPTED or REJECTED -->
+
+# Application Formalization
+
+## Summary
+It is proposed that an Application be defined as the minimal but complete set of elements which together are intended to deliver a specific solution to Folio. The intention is to group together all the closely related components which will need to evolve together, but independently of other parts (i.e. other applications) of Folio.
+
+## Motivation
+The existing Folio project lacks a formal definition of an Application structure. A consequence is that Folio is stuck in a mode of monolithic releases which are ever increasing in size.  There have also been repeated complaints that the setup and management of Folio can be difficult due to the number modules involved.
+
+## Scope
+* Application composition
+* Application versioning
+* Application dependencies
+* Application definition/descriptors
+* Backward compatibility with Folio instances not adopting Applications
+* Application management and deployment are ***out of scope***
+* Application Stores and Marketplaces are ***out of scope***
+* Application-specific UI bundles is ***out of scope***
+* How movement of modules between Applications will work is ***out of scope***
+* How bounded contexts and cross-module database access could be introduced is ***out of scope***
+
+## Detailed Explanation/Design
+
+### Terminology
+* Application:  The minimal but complete set of elements which together are intended to deliver a specific solution to Folio.
+* Bounded context:  The boundary of the microservice.  It "owns" the storage for the service.
+
+### Specifics
+
+#### Application Composition
+Applications can be thought of as vertical slices of functionality.  As such, they're are comprised of a combination of the following:
+* Backend Modules 
+* Edge Modules 
+* UI Modules 
+* UI Plugins 
+
+Applications will typically be made up of multiple components spanning all or several of these categories.  However, that's not a requirement.  In fact, in it's simplest form, it's possible for an application to be comprised of a single module from any of the categories listed above.
+
+An important concept introduced by the formalization of Applications, is that Applications become the delivery mechanism for the modules they contain. Therefore, if a new version of a particular module is required, then it follows that the new version of the entire Application must be installed. It then also follows that a Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.  It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.
```

Comment:
> ... It then also follows that a Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.

I do not think this follows, though I agree that to do otherwise has issues. Two more-specific comments:
* Without the definition of a context (say a particular installation/tenant - not defined here) then this is an unenforceable restriction unless it happens within a particular dependency chain of application descriptors (defined here).
* The current text with unspecified context (which might extend to all versions of all application descriptors) doesn't really make sense and will almost certainly be untrue as applications evolve and modules are moved about.

Because this RFC really only looks in detail at applications and modules I think it might be best to state a looser expectation that could be refined in the remaining RFCs for application formalization that consider composing applications for a tenant. Maybe something like:
> It is expected that a Module will belong to one and only one Application within a given tenant configuration.

Reply:
Do the following adjustments help?
> It then also follows that a **given version of a** Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.  To do otherwise is possible, but would quickly lead to dependency and version management challenges.  **This restriction is enforced by the Application Manager and/or Folio Application Registry (FAR), both of which are out of scope for this RFC.**  It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.

Reply:
I think you still need to specify a context for this make sense (and "follows" is not true though not so important).

Reply:
I think from comments such as https://github.com/folio-org/rfcs/pull/29#discussion_r1605111491 that the intended context is the configuration for a tenant

Reply:
I've incorporated the language above.